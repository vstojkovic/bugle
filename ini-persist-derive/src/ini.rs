use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::spanned::Spanned;
use syn::{Data, DataStruct, DeriveInput, Error, Fields, Ident, Result, Type};

use crate::attr::{IniAttr, NoAttrSupport};

mod attr;
pub mod load;
pub mod save;

use self::attr::FieldAttr;

type FieldExpander = fn(name: &Ident, typ: &Type, section: TokenStream, span: Span) -> TokenStream;
type TraitExpander = fn(struct_name: &Ident, field_expansions: Vec<TokenStream>) -> TokenStream;

fn expand_ini_impl(
    input: DeriveInput,
    field_expander: FieldExpander,
    trait_expander: TraitExpander,
) -> Result<TokenStream> {
    let fields = match &input.data {
        Data::Struct(DataStruct {
            fields: Fields::Named(fields),
            ..
        }) => &fields.named,
        _ => panic!("IniLoad can only be derived on a struct with named fields"),
    };

    NoAttrSupport::from_ast(input.attrs.iter())?;

    let struct_name = input.ident;
    let field_expansions = fields
        .into_iter()
        .map(|field| {
            let attr = FieldAttr::from_ast(field.attrs.iter())?;
            let field_name = field.ident.as_ref().unwrap();
            let span = field.ty.span();

            let section_name = attr.section.unwrap_or_else(|| Some(field_name.to_string()));
            let section = match section_name {
                Some(name) => quote!(Some(#name.to_string())),
                None => quote!(None::<String>),
            };

            Ok(field_expander(field_name, &field.ty, section, span))
        })
        .map(|result| result.unwrap_or_else(Error::into_compile_error))
        .collect();

    Ok(trait_expander(&struct_name, field_expansions))
}
