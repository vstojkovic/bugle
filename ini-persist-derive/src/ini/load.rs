use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{DeriveInput, Ident, Result, Type};

use super::expand_ini_impl;

pub fn expand_ini_load(input: DeriveInput) -> Result<TokenStream> {
    expand_ini_impl(input, expand_field, expand_trait)
}

fn expand_field(name: &Ident, _typ: &Type, section: TokenStream, span: Span) -> TokenStream {
    quote_spanned! { span =>
        if let Some(section) = ini.section(#section) {
            self.#name.load_in(section, "")?;
        }
    }
}

fn expand_trait(struct_name: &Ident, field_expansions: Vec<TokenStream>) -> TokenStream {
    quote! {
        #[automatically_derived]
        impl ini_persist::load::IniLoad for #struct_name {
            fn load_from_ini(&mut self, ini: &ini::Ini) -> ini_persist::Result<()> {
                use ini_persist::load::LoadProperty;
                #(#field_expansions)*
                ini_persist::Result::Ok(())
            }
        }
    }
}
