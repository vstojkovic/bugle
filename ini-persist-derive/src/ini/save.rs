use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{DeriveInput, Ident, Result, Type};

use super::expand_ini_impl;

pub fn expand_ini_save(input: DeriveInput) -> Result<TokenStream> {
    expand_ini_impl(input, expand_field, expand_trait)
}

fn expand_field(name: &Ident, typ: &Type, section: TokenStream, span: Span) -> TokenStream {
    quote_spanned! { span =>
        {
            let section = ini.entry(#section).or_insert_with(ini::Properties::default);
            #typ::remove(section, "");
            self.#name.append(section, "");
        }
    }
}

fn expand_trait(struct_name: &Ident, field_expansions: Vec<TokenStream>) -> TokenStream {
    quote! {
        #[automatically_derived]
        impl ini_persist::save::IniSave for #struct_name {
            fn save_to_ini(&self, ini: &mut ini::Ini) {
                use ini_persist::save::SaveProperty;
                #(#field_expansions)*
            }
        }
    }
}
