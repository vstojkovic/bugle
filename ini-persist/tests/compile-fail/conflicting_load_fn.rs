use ini::Properties;
use ini_persist::{LoadProperty, Result};

#[derive(LoadProperty)]
struct Foo {
    #[ini(load_in_with = my_load_in)]
    #[ini(load_in_with = my_load_in)]
    argle: u8,

    #[ini(load_in_with = my_load_in)]
    #[ini(load_with = my_load)]
    bargle: u8,

    #[ini(load_in_with = my_load_in)]
    #[ini(parse_with = my_parse)]
    glop_glyf: u8,
}

fn my_load_in(field: &mut u8, section: &Properties, key: &str) -> Result<()> {
    todo!()
}

fn my_load(section: &Properties, key: &str) -> Result<Option<u8>> {
    todo!()
}

fn my_parse(value: &str) -> Result<u8> {
    todo!()
}

fn main() {
}
