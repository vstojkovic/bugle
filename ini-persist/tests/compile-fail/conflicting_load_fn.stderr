error: conflicting load function specified
 --> tests/compile-fail/conflicting_load_fn.rs:8:11
  |
8 |     #[ini(load_in_with = my_load_in)]
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^

error: conflicting load function specified
  --> tests/compile-fail/conflicting_load_fn.rs:12:11
   |
12 |     #[ini(load_with = my_load)]
   |           ^^^^^^^^^^^^^^^^^^^

error: conflicting load function specified
  --> tests/compile-fail/conflicting_load_fn.rs:16:11
   |
16 |     #[ini(parse_with = my_parse)]
   |           ^^^^^^^^^^^^^^^^^^^^^
